modules/multisignatures.js

```diff
@@ -80,158 +80,221 @@ function Multisignatures(cb, scope) {
- * @param {function} cb - Callback function
- * @returns {setImmediateCallback} cb, err
- * @todo Add test coverage.
+ * @private
+ * @param {Object} signature - Signature data
+ * @param {string} [signature.publicKey] - Public key of account that created the signature (optional)
+ * @param {string} signature.transactionId - Id of transaction that signature was created for
+ * @param {string} signature.signature - Actual signature
+ * @param {Array} members - Array of multisignature account members - public keys
+ * @param {Object} transaction - Corresponding transaction grabbed from transaction pool
+ * @returns {boolean} isValid - true if signature passed verification, false otherwise
  */
-Multisignatures.prototype.processSignature = function(transaction, cb) {
-	if (!transaction) {
-		return setImmediate(
-			cb,
-			'Unable to process signature. Signature is undefined.'
-		);
-	}
-	const multisignatureTransaction = modules.transactions.getMultisignatureTransaction(
-		transaction.transactionId
-	);
-
-	function done(cb) {
-		library.balancesSequence.add(cb => {
-			const multisignatureTransaction = modules.transactions.getMultisignatureTransaction(
-				transaction.transactionId
-			);
-
-			if (!multisignatureTransaction) {
-				return setImmediate(cb, 'Transaction not found');
+__private.isValidSignature = (signature, members, transaction) => {
+	let isValid = false;
+
+	try {
+		// If publicKey is provided we can perform direct verify
+		if (signature.publicKey) {
+			// Check if publicKey is present as member of multisignature account in transaction
+			if (!members.includes(signature.publicKey)) {
+				library.logger.error(
+					'Unable to process signature, signer not in keysgroup.',
+					{ signature, members, transaction }
+				);
+				return isValid;
 			}
 
-			modules.accounts.getAccount(
-				{
-					address: multisignatureTransaction.senderId,
-				},
-				(err, sender) => {
-					if (err) {
-						return setImmediate(cb, err);
-					} else if (!sender) {
-						return setImmediate(cb, 'Sender not found');
-					}
-					multisignatureTransaction.signatures =
-						multisignatureTransaction.signatures || [];
-					multisignatureTransaction.signatures.push(transaction.signature);
-					multisignatureTransaction.ready = Multisignature.prototype.ready(
-						multisignatureTransaction,
-						sender
-					);
-
-					library.bus.message('signature', transaction, true);
-					return setImmediate(cb);
-				}
+			// Try to verify the signature
+			isValid = library.logic.transaction.verifySignature(
+				transaction,
+				signature.publicKey,
+				signature.signature
+			);
+		} else {
+			// publicKey is not there - we need to iterate through all members of multisignature account in transaction
+			isValid = members.find(member =>
+				// Try to verify the signature
+				library.logic.transaction.verifySignature(
+					transaction,
+					member,
+					signature.signature
+				)
```

Comment:
This code block will assign a `publicKey` to `isValid`, while its expected to be boolean value. If later on use someone did a strict match with `true/false` then code will break.